feat: support for Ruuvi Air data format E1

rotvalli opened the pull request:
Ruuvi Air E1 format decoder based on work done on dfe1_decoder:
https://github.com/Bluetooth-Devices/ruuvitag-ble/blob/main/src/ruuvitag_ble/dfe1_decoder.py
And specs by Ruuvi Innovations Ltd:
https://docs.ruuvi.com/communication/bluetooth-advertisements/data-format-e1.md

ttu commented on ruuvitag_sensor/data_formats.py:
```diff
@@ -92,6 +92,10 @@ def convert_data(raw: str) -> DataFormatAndRawSensorData:  # noqa: PLR0911
         log.debug("Found candidate %s", candidate)
 
+        # Ruuvi Air (E1)
+        if candidate.startswith("FF990406"):
```
I think extended might have different value (FF9904E1) based on the documentation?
https://docs.ruuvi.com/communication/bluetooth-advertisements/data-format-e1

ttu commented on ruuvitag_sensor/decoder.py:
```diff
@@ -283,6 +286,133 @@ def decode_data(self, data: str) -> Optional[SensorData5]:
+    def _get_mac(self, data: ByteData) -> str:
+        return ":".join(f"{b:02X}" for b in bytes(data[16]))
+
+    def decode_data(self, data: str) -> Optional[SensorData3]:
```
This should return new SensorDataE1

rotvalli replied:
Fiexd

ttu commented:
Thanks, this addition looks good. I implemented the support for data format 6 in https://github.com/ttu/ruuvitag-sensor/pull/279. Sorry for the conflicts.

 It seems that Ruuvi Air broadcasts both FF9904e1 and FF990406. Documentation says that if E1 is sent, then DF6 should be discarded. Maybe it is not the responsibility of this package to make the discarding decision, so the implementation will be more simple.
```sh
Found candidate FF99040511d6545bc61affccfffc03f4a9b6214519caf744deebe1
Found candidate FF9904e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034ecb8ffffffffffd9380eab71d4
Found candidate FF9904e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034edb8ffffffffffd9380eab71d4
Found candidate FF990406116f48d0c6350013031c0f00ffffef90ab71d4e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034edb8ffffffffffd9380eab71d4
Found candidate FF990406116f48d0c6350013031c0f00ffffef90ab71d4e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034edb8ffffffffffd9380eab71d4
Found candidate FF990406116f48d0c6350013031c0f00fffff090ab71d4e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034edb8ffffffffffd9380eab71d4
Found candidate FF990406116f48d0c6350013031c0f00fffff090ab71d4e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034edb8ffffffffffd9380eab71d4
```

ojousima commented:
> It seems that Ruuvi Air broadcasts both FF9904e1 and FF990406. Documentation says that if E1 is sent, then DF6 should be discarded. Maybe it is not the responsibility of this package to make the discarding decision, so the implementation will be more simple.

Yes, Air sends always both. 

Format 6 uses BLE4 advertising and is included for compatibility with all BLE4 devices. 

Format E1 has more data and is preferred, but it uses BLE5 advertising extensions which are not seen by devices which do not have BLE5 support. 

I'll clarify the wording on documentation.

ttu commented on ruuvitag_sensor/data_formats.py:
```diff
@@ -103,6 +104,9 @@ def convert_data(raw: str) -> DataFormatAndRawSensorData:  # noqa: PLR0911
             return (6, candidate[6:])
 
+        if candidate.startswith("FF9904E1"):
+            return ("E1", candidate[6:])
```
As this returns DataFormatAndRawSensorData, we need to change DataFormat  to support also str

https://github.com/ttu/ruuvitag-sensor/blob/62ee2f6c3d8a145e030c25462fb8547d89252cb3/ruuvitag_sensor/ruuvi_types.py#L65

ttu commented on ruuvitag_sensor/decoder.py:
```diff
@@ -294,6 +297,133 @@ def decode_data(self, data: str) -> Optional[SensorData5]:
+        try:
+            byte_data: ByteData = struct.unpack(">BhHHHHHHHBB3s3s3sB5s6s", bytearray.fromhex(data[:80]))
+            return {
+                "data_format": "E1",
```
as data format can now be also string, we need to change data_format to be int or str
https://github.com/ttu/ruuvitag-sensor/blob/62ee2f6c3d8a145e030c25462fb8547d89252cb3/ruuvitag_sensor/ruuvi_types.py#L5

ttu commented on ruuvitag_sensor/decoder.py:
```diff
@@ -294,6 +297,133 @@ def decode_data(self, data: str) -> Optional[SensorData5]:
+    def _get_mac(self, data: ByteData) -> str:
+        return ":".join(f"{b:02X}" for b in bytes(data[16]))
+
+    def decode_data(self, data: str) -> Optional[SensorDataE1]:
```
class for SensorDataE1 is missing. As it extends data format 6, maybe you can inherit it from this class.
https://github.com/ttu/ruuvitag-sensor/blob/62ee2f6c3d8a145e030c25462fb8547d89252cb3/ruuvitag_sensor/ruuvi_types.py#L42

ttu commented on ruuvitag_sensor/data_formats.py:
```diff
@@ -103,6 +103,9 @@ def convert_data(raw: str) -> DataFormatAndRawSensorData:  # noqa: PLR0911
         if candidate.startswith("FF990406"):
             return (6, candidate[6:])
 
+        if candidate.startswith("FF9904E1"):
```
Received data had e1 in lowercase (`FF9904e1116f....`). I can check in a separate PR how to make all these checks in the whole codebase case insensitive.
```
Found candidate FF9904e1116f48d0c63400090013001a001e031c0f00ffffffffffff0034ecb8ffffffffffd9380eab71d4
```

ttu commented on tests/test_data_formats.py:
```diff
@@ -42,6 +61,15 @@ def test_get_data_format_5_valid_data(self):
             encoded = DataFormats._get_data_format_5(x)
             assert encoded is not None
 
+    def test_get_data_format_e1_valid_data(self):
```
You could add similar teses that were added for data format 6 in here https://github.com/ttu/ruuvitag-sensor/commit/62ee2f6c3d8a145e030c25462fb8547d89252cb3

You can check cases from official documentation: https://docs.ruuvi.com/communication/bluetooth-advertisements/data-format-e1

ttu commented:
@rotvalli can you add one more test for decoding the data? Similar as `test_df6_decode_valid_data`. Official documentation has good data examples (https://docs.ruuvi.com/communication/bluetooth-advertisements/data-format-e1).

ttu commented on ruuvitag_sensor/data_formats.py:
```diff
@@ -103,6 +103,9 @@ def convert_data(raw: str) -> DataFormatAndRawSensorData:  # noqa: PLR0911
         if candidate.startswith("FF990406"):
             return (6, candidate[6:])
 
+        if candidate.lower().startswith("FF9904E1".lower()):
```
You can add here a comment why we use lowercase for comparisons.
```
# Ruuvi Air sends E1 in lowercase, e.g. `FF9904e1...`
```

Also this `"FF9904E1".lower()` can be `"ff9904e1"`

rotvalli replied:
Thanks for verifying that. Added case insensitive format checking back.

rotvalli commented:
> @rotvalli can you add one more test for decoding the data? Similar as `test_df6_decode_valid_data`. Official documentation has good data examples (https://docs.ruuvi.com/communication/bluetooth-advertisements/data-format-e1).

Added E1 format decoder tests and changed VOC and NOX code to use more readable code from format 6 in the decoder

ttu commented:
Everything looks good now. Thanks for the implementation!
